languages/csharp/Bitwarden.Sdk/BitwardenLibrary.cs

```diff
@@ -2,20 +2,54 @@
 
 namespace Bitwarden.Sdk;
 
-internal static class BitwardenLibrary
+internal static partial class BitwardenLibrary
 {
-    [DllImport("bitwarden_c", CallingConvention = CallingConvention.Cdecl)]
-    private static extern BitwardenSafeHandle init(string settings);
+    [LibraryImport("bitwarden_c", StringMarshalling = StringMarshalling.Utf8)]
+    private static partial BitwardenSafeHandle init(string settings);
 
-    [DllImport("bitwarden_c", CallingConvention = CallingConvention.Cdecl)]
-    private static extern void free_mem(IntPtr handle);
+    [LibraryImport("bitwarden_c", StringMarshalling = StringMarshalling.Utf8)]
+    private static partial void free_mem(IntPtr handle);
 
-    [DllImport("bitwarden_c", CallingConvention = CallingConvention.Cdecl)]
-    private static extern string run_command(string json, BitwardenSafeHandle handle);
+    [LibraryImport("bitwarden_c", StringMarshalling = StringMarshalling.Utf8)]
+    private static partial string run_command(string json, BitwardenSafeHandle handle);
+
+    internal delegate void OnCompleteCallback(IntPtr json);
+
+    [LibraryImport("bitwarden_c", StringMarshalling = StringMarshalling.Utf8)]
+    private static partial void run_command_async(string json, BitwardenSafeHandle handle, OnCompleteCallback on_completed_callback);
 
     internal static BitwardenSafeHandle Init(string settings) => init(settings);
 
     internal static void FreeMemory(IntPtr handle) => free_mem(handle);
 
     internal static string RunCommand(string json, BitwardenSafeHandle handle) => run_command(json, handle);
+
+    internal static Task<string> RunCommandAsync(string json, BitwardenSafeHandle handle, CancellationToken token = default)
+    {
+        var tcs = new TaskCompletionSource<string>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+        try
+        {
+            run_command_async(json, handle, (resultPointer) =>
+            {
+                try
+                {
+                    var stringResult = Marshal.PtrToStringUTF8(resultPointer);
+                    tcs.SetResult(stringResult);
+                }
+                finally
+                {
+                    Marshal.FreeCoTaskMem(resultPointer);
```

Comment:
Is this freeing the Rust created result string? I don't think that's allowed according to the Rust docs, we'd need to free it from the Rust side:

https://doc.rust-lang.org/std/ffi/struct.CString.html#method.into_raw

Instead we should be freeing it from Rust after the callback was executed

Reply:
Good call, moved the freeing to rust.